File: joss/comparisons/cat-5-schema-composition/06-schema-validation/example-schema-validation.py
"""
Error Class 6: Schema Validation at Data Boundaries

Python/pandas reads CSV files with inferred types. There is no upfront
schema validation. Type issues, missing columns, and malformed data
are discovered gradually — sometimes silently, sometimes much later.
"""
import pandas as pd

# ── PROBLEM 6a: No schema validation at load ───────────────────────────
# pandas infers types. If result_value has a non-numeric value like "pending",
# the entire column becomes object (string) type — silently.
labs = pd.read_csv("fixtures/lab_results.csv")

# You can add dtype specifications, but they error at runtime, not compile time
# labs = pd.read_csv("fixtures/lab_results.csv", dtype={"result_value": float})
# ValueError at runtime if any value can't be converted — but only then

# ── PROBLEM 6b: Missing columns discovered late ────────────────────────
# If the CSV is missing a column, you won't know until you access it.
# No upfront validation.
try:
    print(labs["nonexistent_column"])
except KeyError:
    print("KeyError — discovered only when accessed, not at load time")

# ── PROBLEM 6c: NaN for missing required values ────────────────────────
# Empty cells become NaN. There's no way to declare "this column must not
# have missing values" at load time.
print(labs["result_value"].dtype)  # float64 — but could contain NaN
print(f"Null count: {labs['result_value'].isna().sum()}")
# No error even if nulls exist in a column that should be non-null

# ── PROBLEM 6d: No compile-time tracking ───────────────────────────────
# After loading, any column reference is unchecked until runtime.
# Typos, wrong column names, removed columns — all runtime errors.
filtered = labs[labs["reslt_value"] > 100]  # Typo — KeyError at runtime

File: joss/comparisons/cat-5-schema-composition/06-schema-validation/example-schema-validation.R
# Error Class 6: Schema Validation at Data Boundaries
#
# R/readr infers types from the first 1000 rows. There is no upfront
# schema contract. Type mismatches are warnings, not errors, and missing
# columns are only discovered when accessed.

library(tidyverse)

# ── PROBLEM 6a: Type inference, not validation ──────────────────────────
# read_csv guesses column types from the first 1000 rows.
# If row 1001 has "pending" in result_value, the column becomes character.
# Earlier rows' numbers become strings — silently.
labs <- read_csv("fixtures/lab_results.csv")

# You can specify col_types, but mismatches produce warnings, not errors:
# labs <- read_csv("fixtures/lab_results.csv", col_types = cols(
#   result_value = col_double()
# ))
# Parsing failures produce a warning and the offending values become NA

# ── PROBLEM 6b: Missing columns discovered late ────────────────────────
# If the CSV is missing a column, read_csv succeeds.
# You only find out when you try to use it.
# labs %>% mutate(x = nonexistent_column)
# Error: object 'nonexistent_column' not found — runtime only

# ── PROBLEM 6c: NA for missing required values ─────────────────────────
# Empty cells become NA. There's no way to declare "this column must
# not have NA" at load time.
summary(labs$result_value)  # Shows NA count, but doesn't prevent them

# ── PROBLEM 6d: No compile-time tracking ───────────────────────────────
# After loading, all column references are checked at runtime only.
labs %>%
  filter(reslt_value > 100)  # Typo — runtime error
